[home][ios] restore QR code functionality to unrestricted IOS client builds

quinlanj opened the pull request:
# Why

we want to bring back QR code functionality to unrestricted ios client builds (ie) adhoc builds
for reference: 
- https://stackoverflow.com/questions/17584426/check-if-app-is-ad-hocdevapp-store-build-at-run-time 
- https://stackoverflow.com/questions/3426467/how-to-determine-at-run-time-if-app-is-for-development-app-store-or-ad-hoc-dist

# How

- determine what type of client build we have at runtime on ios by looking at `embedded.mobileprovision` . client build types include `adhoc`, `appstore`, `simulator`, etc.
- display QR code when we have an unrestricted client build

# Test Plan
- [x] simulator
- [x] development
- [x] adhoc
- [x] appstore
* adhoc and appstore were tested by making the provisioning profiles on my developer account, bundling them in the expo app and changing the mobile provisioning path to look for these test profiles

quinlanj committed:
[home] add QR code functionality back to unrestricted ios builds

quinlanj committed:
[ios] add methods to determine client build type

brentvatne commented on ios/Exponent/Kernel/DevSupport/EXHomeModule.m:
```diff
@@ -21,6 +21,144 @@ @implementation EXHomeModule
 
 + (NSString *)moduleName { return @"ExponentKernel"; }
 
+/** embedded.mobileprovision plist format:
```
useful to have this here 👍

brentvatne commented on ios/Exponent/Kernel/DevSupport/EXHomeModule.m:
```diff
@@ -21,6 +21,144 @@ @implementation EXHomeModule
+    plistString = [NSString stringWithFormat:@"%@</plist>",plistString];
+    // juggle latin1 back to utf-8!
+    NSData *plistdata_latin1 = [plistString dataUsingEncoding:NSISOLatin1StringEncoding];
+    //    plistString = [NSString stringWithUTF8String:[plistdata_latin1 bytes]];
```
are these comments left here intentionally?

quinlanj replied:
no...

brentvatne commented on ios/Exponent/Kernel/DevSupport/EXHomeModule.m:
```diff
@@ -21,6 +21,144 @@ @implementation EXHomeModule
+  }
+}
+
+- (EXClientReleaseType) getClientReleaseTypeAsync
```
this looks the same as `getClientReleaseType` and isn't async, perhaps we should get rid of `getClientReleaseTypeAsync`?

brentvatne commented on ios/Exponent/Kernel/DevSupport/EXHomeModule.m:
```diff
@@ -39,7 +177,8 @@ + (BOOL)requiresMainQueueSetup
 {
-  return @{ @"sdkVersions": _sdkVersions };
+  return @{ @"sdkVersions": _sdkVersions,
+            @"IOSClientReleaseType": [self clientReleaseTypeToJS: [self getClientReleaseTypeAsync]] };
```
```suggestion
            @"IOSClientReleaseType": [self clientReleaseTypeToJS: [self getClientReleaseType]] };
```

maybe this if we remove `getClientReleaseTypeAsync`

brentvatne reviewed (changes requested):
a couple comments, otherwise looks good!

esamelson commented on ios/Exponent/Kernel/DevSupport/EXHomeModule.m:
```diff
@@ -21,6 +21,144 @@ @implementation EXHomeModule
+ */
+
+-(NSDictionary*) getMobileProvision {
+  static NSDictionary* mobileProvision = nil;
```
since this method is an instance method, i'd use a property in the class to save the value instead of making a static variable. you can see an example of this here https://github.com/expo/expo/blob/5bf4c3d5071cf72469fb41e3dc71ff93d3c5ff79/ios/Exponent/Kernel/AppLoader/CachedResource/EXManifestResource.m#L185-L219 (property defined on L23)

esamelson reviewed (changes requested):
few obj-c nits but looks good overall, thanks for figuring out how to do this with the mobileprovision file! 😮

quinlanj committed:
pr feedback (brent)

quinlanj committed:
pr feedback (eric)

quinlanj committed:
add client release type header to HTTP requests for www permissions checking

quinlanj commented:
sorry for the last minute changes but here are some additional things i added to make qrcodes work again:
- added headers requests sent in EXFileDownloader so we can let `www` know to give us the manifest if the client is not in the appstore

esamelson commented on ios/Exponent/Kernel/AppLoader/EXFileDownloader.m:
```diff
@@ -1,6 +1,7 @@
 // Copyright 2015-present 650 Industries. All rights reserved.
 
 #import "EXEnvironment.h"
+#import "EXHomeModule.h"
```
would be great to put these methods in a different module like `EXProvisioningProfile` so you don't have to import EXHomeModule here.

esamelson reviewed (approved):
sorry for the long feedback chain! lgtm other than moving stuff to `EXProvisioningProfile`.

quinlanj committed:
[ios] use the existing provisioningprofile parser instead

quinlanj committed:
[ios] fix build type logic to use new parser method
